### PR TITLE
i#3990: Initialises raw tls slots to zero for Linux

### PR DESCRIPTION
**Pull request by johnfxgalea**
Removes the union in unix/tls.h to separate os_seg_info from client slots. As a result, client slots are correctly initialised to zero as specified in the docs.

Adds a regression test to ensure slots are initialised correctly.

Issue: #3990

### CODE REVIEW

File: core/unix/tls.h [hunk 1]
```diff
@@ -263,6 +263,15 @@ typedef struct _os_local_state_t {
 #endif
     void *app_lib_tls_base; /* for mangling segmented memory ref */
     void *app_alt_tls_base; /* for mangling segmented memory ref */
+
+/* FIXME i#3990: For MACOS, we use a union to save tls space. Unfortunately, this
+ * results in not initialising client tls slots which are allocated using
+ * dr_raw_tls_calloc. Figuring where to perform memset to clear os_seg_info is not
+ * apparently clear due to interleaved thread and instrum inits.  */
```
Comment (derekbruening) [inline core/unix/tls.h]: style nit: prefer `*/` on own line.

File: suite/tests/client-interface/raw_tls.dll.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+static void **
+get_tls_addr(int slot_idx)
+{
+    byte *base = dr_get_dr_segment_base(tls_raw_reg);
+    byte *addr = (byte *)(base + tls_raw_base + slot_idx * sizeof(void *));
+    return *((void **)addr);
+}
+
+static void
+check()
```
Comment (derekbruening) [inline suite/tests/client-interface/raw_tls.dll.c]: nit: C needs `(void)` to ensure no args.

**Comment by derekbruening** [comment]
> I#3990: Initialises raw tls slots to zero for Linux

nit: s/I/i/

File: suite/tests/client-interface/raw_tls.dll.c (hunk 2 above)
Comment (johnfxgalea) [reply suite/tests/client-interface/raw_tls.dll.c]: Done.

File: core/unix/tls.h (hunk 1 above)
Comment (johnfxgalea) [reply core/unix/tls.h]: Done

**Comment by johnfxgalea** [comment]
Thanks for taking a look.

**Comment by johnfxgalea** [comment]
Are these CI failures related?

**Comment by derekbruening** [comment]
Traveling. Did you search the tracker. Look at recent PRs: some tests that
need attention.

**Comment by johnfxgalea** [comment]
> Traveling. Did you search the tracker. Look at recent PRs: some tests that need attention.
> […](#)

Yes I did check the tracker, and other recent PRs, which failed for similar reasons. 

We can address this after the holidays; I am not in a rush.

**Comment by derekbruening** [comment]
I would suggest listing the failed tests and issue numbers here.  If all are known we can merge this.  Hopefully someone will get time to improve the CI.

**Comment by johnfxgalea** [comment]
> I would suggest listing the failed tests and issue numbers here. If all are known we can merge this. Hopefully someone will get time to improve the CI.

**Travis**

> ====> FAILURE in debug-internal-64 <====
> 1692debug-internal-64: 328 tests passed, **** 3 tests failed, of which 1 were flaky, but ignoring 0 for i#2941: ****
> 1693	code_api|client.drmgr-test -------------  **`#3496`** 
> 1694	code_api|tool.drcacheoff.burst_malloc ---------- **`#3974`** 
> 1695	code_api|api.static_sideline_FLAKY 

**AppVeyor**

> 26068debug-internal-32: 258 tests passed, **** 2 tests failed, of which 1 were flaky: ****
> 26069	code_api|security-win32.codemod-threads_FLAKY 
> 26070	code_api|client.pcache-use ---------- **`#2145`**
> 26071debug-internal-64: 228 tests passed, **** 4 tests failed: ****
> 26072	code_api|client.loader 
> 26073	code_api|client.thread ----------- **`#3560`**
> 26074	code_api|client.nudge_ex 
> 26075	code_api|tool.drcachesim.coherence ---------- **`#3971`**          

Failures which I did not map back to their corresponding issue numbers are either cause by flaky tests or are ignored by the CI itself.

Having said that, I think we should still try to improve the CI. LMK if I can help, especially if you already have a fix in mind.

**Comment by derekbruening** [comment]
> Failures which I did not map back to their corresponding issue numbers are either caused by flaky tests or are ignored by the CI itself.

What about client.loader and client.nudge_ex?  Let's get issues filed on those 2 with the failure details (assert?  output mismatch?).  Best to not have any unaccounted failures.

**Comment by johnfxgalea** [comment]
```
debug-internal-64: 228 tests passed, **** 4 tests failed, but ignoring 2 for i#2145: ****
[00:32:51] 	(ignore: i#2145) 	code_api|client.loader 
[00:32:51] 	code_api|client.thread 
[00:32:51] 	(ignore: i#2145) 	code_api|client.nudge_ex 
[00:32:51] 	code_api|tool.drcachesim.coherence 
```

The CI is ignoring those and referring to issue #2145

**Comment by derekbruening** [comment]
OK.  If you can repro any of these (besides #3974 which is on my plate) and submit a fix that would be appreciated.

**Pull request merged**